compiler/noirc_frontend/src/elaborator/expressions.rs

```diff
@@ -230,6 +236,41 @@ impl Elaborator<'_> {
         (HirBlockExpression { statements }, block_type)
     }
 
+    /// If the given type was declared as:
+    /// - `#[must_use = "message"]`, return `Some(Some(message))`
+    /// - `#[must_use]`, return `Some(None)`
+    /// - otherwise, return `None`
+    fn type_is_must_use(typ: &Type) -> MustUse {
+        /// Helper function to avoid infinite recursion for infinitely recursive types
```

Comment:
Would we get an additional error for a self referential type that it must be used? I don't think we would based upon the elaboration ordering (we check for self referential types when resolving named types) but it would be good to have a unit test for this case.